cns/service.go

```diff
@@ -48,63 +49,79 @@ func NewService(name, version, channelMode string, store store.KeyValueStore) (*
+	// only use TLS connection for DNC/CNS listener:
+	if config.TlsSettings.TLSPort != "" {
+		// listener.URL.Host will always be hostname:port, passed in to CNS via CNS command
+		// else it will default to localhost
+		// extract hostname and override tls port.
+		hostParts := strings.Split(nodeListener.URL.Host, ":")
+		tlsAddress := net.JoinHostPort(hostParts[0], config.TlsSettings.TLSPort)
+
+		// Start the listener and HTTP and HTTPS server.
+		tlsConfig, err := getTLSConfig(config.TlsSettings, config.ErrChan) //nolint
+		if err != nil {
+			log.Printf("Failed to compose Tls Configuration with error: %+v", err)
+			return errors.Wrap(err, "could not get tls config")
+		}
+
+		if err := nodeListener.StartTLS(config.ErrChan, tlsConfig, tlsAddress); err != nil {
+			return errors.Wrap(err, "could not start tls")
+		}
+	}
+
+	nodeListener.ListenerType = "nodeListener"
+	*config.Listeners = append(*config.Listeners, *nodeListener)
+
+	// bind on localhost ip for CNI listener
+	localURL, _ := url.Parse(defaultAPIServerURL)
+	localListener, err := acn.NewListener(localURL)
+	if err != nil {
+		return errors.Wrap(err, "Failed to construct url for local listener")
+	}
+
+	localListener.ListenerType = "localListener"
+	*config.Listeners = append(*config.Listeners, *localListener)
+
+	logger.Printf("HTTP listeners will be started later after CNS state has been reconciled")
+	*service.Listeners = *config.Listeners
+
+	return nil
 }
 
 // Initialize initializes the service and starts the listener.
-func (service *Service) Initialize(config *common.ServiceConfig) error {
+func (service *Service) Initialize(config *common.ServiceConfig, primaryIP string) error {
 	log.Debugf("[Azure CNS] Going to initialize a service with config: %+v", config)
 
 	// Initialize the base service.
 	if err := service.Service.Initialize(config); err != nil {
 		return errors.Wrap(err, "failed to initialize")
 	}
 
-	// Initialize the listener.
-	if config.Listener == nil {
-		// Fetch and parse the API server URL.
-		u, err := url.Parse(service.getAPIServerURL())
-		if err != nil {
-			return err
-		}
-
-		listener, err := acn.NewListener(u)
-		if err != nil {
-			return err
-		}
-
-		if config.TlsSettings.TLSPort != "" {
-			// listener.URL.Host will always be hostname:port, passed in to CNS via CNS command
-			// else it will default to localhost
-			// extract hostname and override tls port.
-			hostParts := strings.Split(listener.URL.Host, ":")
-			tlsAddress := net.JoinHostPort(hostParts[0], config.TlsSettings.TLSPort)
-
-			// Start the listener and HTTP and HTTPS server.
-			tlsConfig, err := getTLSConfig(config.TlsSettings, config.ErrChan)
-			if err != nil {
-				log.Printf("Failed to compose Tls Configuration with error: %+v", err)
-				return errors.Wrap(err, "could not get tls config")
-			}
-
-			if err := listener.StartTLS(config.ErrChan, tlsConfig, tlsAddress); err != nil {
-				return err
-			}
+	// Initialize listeners.
+	if config.Listeners == nil {
```

Comment:
you should check for whether `config.Listeners` is empty, not nil, since it is a list